src/repl.ml

```diff
@@ -0,0 +1,191 @@
+       build it. *)
+    Promise.return false
+  | el :: _ -> (
+    let cwd = Filename.dirname (Uri.fsPath el) |> Path.of_string in
+    let cmd = Sandbox.get_command sandbox "dune" [ "build" ] in
+    let+ result = Cmd.output ~cwd cmd in
+    match result with
+    | Error _ -> false
+    | Ok _ -> true )
+
+let default_repl sandbox =
+  let open Promise.Syntax in
+  let* has_utop = has_utop sandbox in
+  let+ can_build = can_build sandbox in
+  match (has_utop, can_build) with
+  | true, true -> Sandbox.get_command sandbox "dune" [ "utop" ]
+  | true, false -> Sandbox.get_command sandbox "utop" []
+  | false, _ -> Sandbox.get_command sandbox "ocaml" []
+
+let create_terminal instance sandbox =
+  match Extension_instance.repl instance with
+  | Some term ->
+    Terminal_sandbox.show term;
+    Promise.return (Ok term)
+  | None -> (
+    let open Promise.Syntax in
+    let* cmd =
+      match (get_repl_path (), get_repl_args ()) with
+      | Some bin, Some args ->
+        Sandbox.get_command sandbox bin args |> Promise.return
+      | Some bin, None -> Sandbox.get_command sandbox bin [] |> Promise.return
+      | None, _ -> default_repl sandbox
+    in
+    let* result = Cmd.check cmd in
+    match result with
+    | Error err -> Promise.return (Error err)
+    | Ok (Shell _) ->
+      Promise.return
+        (Error "Running a REPL from a Shell command is not supported")
+    | Ok (Spawn cmd) -> (
+      let term = Terminal_sandbox.create ~name ~command:cmd sandbox in
+      Terminal_sandbox.show term;
+      (* Wait for UTop or OCaml REPL to be initialized before sending text to
+         the terminal.
+
+         That's hacky, buy hey, if vscode-python does it, so can we...
+         https://github.com/microsoft/vscode-python/blob/main/src/client/terminals/codeExecution/terminalCodeExecution.ts#L54 *)
+      let+ _ =
+        Promise.make (fun ~resolve ~reject:_ ->
+            let (_ : Node.Timeout.t) =
+              Node.setTimeout (fun () -> resolve true) 2500
+            in
+            ())
+      in
+      match Terminal.exitStatus term with
+      | Some _ -> Error "The REPL terminal could not be open"
+      | None ->
+        Extension_instance.set_repl instance term;
+        Ok term ) )
+
+let get_code text_editor =
+  let selection = Vscode.TextEditor.selection text_editor in
+  let start_line = Vscode.Selection.start selection |> Vscode.Position.line in
+  let end_line = Vscode.Selection.end_ selection |> Vscode.Position.line in
+  let start_char =
+    Vscode.Selection.start selection |> Vscode.Position.character
+  in
+  let end_char = Vscode.Selection.end_ selection |> Vscode.Position.character in
+  if start_line = end_line && start_char = end_char then
+    let document = Vscode.TextEditor.document text_editor in
+    let line = Vscode.TextDocument.lineAt document ~line:start_line in
+    Vscode.TextLine.text line
+  else
+    let document = Vscode.TextEditor.document text_editor in
```

Comment:
This binding occurs in both `then` and `else` clauses. Hence it can exist outside the `if` altogether.